### PR TITLE
OSD-15539: Write cluster specific kube config file and enable multi cluster login / logout

### PR DESCRIPTION
**Pull request by samanthajayasinghe**
### What type of PR is this?

feature

### What this PR does / Why we need it?

BP-CLI allows to login/logout multi clusters  
```
ocm backplane login {clusterid-1} --multi 
ocm backplane login {clusterid-2} --multi 
```
### Which Jira/Github issue(s) does this PR fix?

_Resolves #_

[OSD-15539](https://issues.redhat.com//browse/OSD-15539)

### Pre-checks (if applicable)

- [x] Ran unit tests locally
- [x] Validated the changes in a cluster
- [x] Included documentation changes with PR

### CODE REVIEW

**Comment by supreeth7** [comment]
@samanthajayasinghe is this still WIP? If not, can you change the title?

**Comment by samanthajayasinghe** [comment]
@supreeth7 Need to do a few things with this PR 
* clear cluster specific config when logout from a specific cluster 
* some tests to cover new functionality 

I'll open the PR after that :)

File: cmd/ocm-backplane/login/login.go [hunk 1]
```diff
@@ -56,6 +57,13 @@ func init() {
 		false,
 		"Login to management cluster instead of the cluster itself.",
 	)
+	flags.BoolVarP(
+		&args.multiCluster,
+		"multi",
+		"m",
+		false,
+		"Enable multi cluster login.",
```
Comment (supreeth7) [inline cmd/ocm-backplane/login/login.go:65]:
```suggestion
		"Enable multi-cluster login.",
```

File: cmd/ocm-backplane/login/login_test.go [hunk 2]
```diff
@@ -168,5 +170,25 @@ var _ = Describe("Login command", func() {
 
 			Expect(err).To(BeNil())
 		})
+
+		It("Check KUBECONFIG when login into multi clusters.", func() {
```
Comment (supreeth7) [inline cmd/ocm-backplane/login/login_test.go:174]:
```suggestion
		It("Check KUBECONFIG when logging into multiple clusters.", func() {
```

File: cmd/ocm-backplane/login/login_test.go [hunk 3]
```diff
@@ -168,5 +170,25 @@ var _ = Describe("Login command", func() {
...
+			args.manager = false
+			args.multiCluster = true
+			_ = utils.ModifyTempKubeConfigFileName(trueClusterId)
```
Comment (supreeth7) [inline cmd/ocm-backplane/login/login_test.go:177]: Why was the error not handled here intentionally?

**Review by supreeth7 — CHANGES_REQUESTED** [review]

**Comment by samanthajayasinghe** [comment]
Hi @supreeth7 
I'll address all comments after adding the proxy support to all BP-CLI commands ( aka [PR](https://github.com/openshift/backplane-cli/pull/46) )

File: cmd/ocm-backplane/login/login_test.go (hunk 3 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/login/login_test.go:177]: Handle the error

File: cmd/ocm-backplane/login/login_test.go (hunk 2 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/login/login_test.go:174]: done

File: cmd/ocm-backplane/login/login.go (hunk 1 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/login/login.go:65]: done

**Comment by samanthajayasinghe** [comment]
Hi @supreeth7 
Can you pls `unhold` this PR and have a look?

File: cmd/ocm-backplane/login/login.go [hunk 4]
```diff
@@ -93,6 +101,9 @@ func runLogin(cmd *cobra.Command, argv []string) (err error) {
 			return err
 		}
 		logger.Debugf("Using backplane Proxy URL: %s\n", proxyUrl)
+	} else {
+
+		proxyUrl = bpConfig.ProxyURL
```
Comment (supreeth7) [inline cmd/ocm-backplane/login/login.go:106]: Isn't this already being done in the next line?
i.e; `if len(proxyUrl) == 0` (which will be the case if `--proxy` isn't specified) then set the bpConfig's `ProxyURL` as `proxyUrl`.

File: cmd/ocm-backplane/logout/logout.go [hunk 5]
```diff
@@ -26,65 +27,74 @@ var LogoutCmd = &cobra.Command{
...
-		return err
+	// Logout specific cluster
+	if len(argv) == 1 {
+		err := login.RemoveClusterKubeConfig(argv[0])
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Logged out from backplane: %s\n", argv[0])
+	} else {
+		rc, err := utils.ReadKubeconfigRaw()
```
Comment (supreeth7) [inline cmd/ocm-backplane/logout/logout.go:38]: Can you please elaborate what `rc` means here?

File: cmd/ocm-backplane/logout/logout.go [hunk 6]
```diff
@@ -26,65 +27,74 @@ var LogoutCmd = &cobra.Command{
...
+		currentClusterObj := rc.Clusters[currentCluster]
+		if currentClusterObj == nil {
+			return fmt.Errorf("current cluster not found, skipping")
+		}
+		currentServer := currentClusterObj.Server
+
+		// backplane should only handle `logout` associated context
+		// created with backplane itself, we check this via matching
+		// the cluster server endpoint
+
```
Comment (supreeth7) [inline cmd/ocm-backplane/logout/logout.go:65]:
```suggestion
```

File: cmd/ocm-backplane/logout/logout.go [hunk 7]
```diff
@@ -26,65 +27,74 @@ var LogoutCmd = &cobra.Command{
...
+		logger.Debugln("Logging out of the current cluster")
+
+		// Delete the current cluster/context/user and set current-context to empty str
+		delete(rc.Clusters, currentCluster)
+		delete(rc.Contexts, rc.CurrentContext)
+		delete(rc.AuthInfos, currentUser)
+		savedContext := rc.CurrentContext
+		// Setting current-context to empty str will make `oc` command throwing
+		// errors saying that config is incomplete, however, this is inline with
+		// the behavior of `oc config unset current-context`
```
Comment (supreeth7) [inline cmd/ocm-backplane/logout/logout.go:87]:
```suggestion
		// Setting current-context to empty str will make `oc` command return
		// errors saying that the config is incomplete, however, this is inline with
		// the behavior of `oc config unset current-context`
```

File: cmd/ocm-backplane/logout/logout.go [hunk 8]
```diff
@@ -26,65 +27,74 @@ var LogoutCmd = &cobra.Command{
...
+		rc.CurrentContext = ""
+
+		pathOptions := clientcmd.NewDefaultPathOptions()
+		err = clientcmd.ModifyConfig(pathOptions, rc, true)
+		if err != nil {
+			return err
+		}
+		logger.Debugln("Wrote kubeconfig")
```
Comment (supreeth7) [inline cmd/ocm-backplane/logout/logout.go:95]:
```suggestion
		logger.Debugln("Kubeconfig written")
```

File: cmd/ocm-backplane/login/login.go (hunk 4 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/login/login.go:106]: yeah .. You are right.

File: cmd/ocm-backplane/logout/logout.go (hunk 5 above)
Comment (samanthajayasinghe) [reply cmd/ocm-backplane/logout/logout.go:38]: `rc` mean ReadConfig from the default config location( ~/.kube/config).
I added a comment to the code.

**Comment by samanthajayasinghe** [comment]
> My $0.2 - While I like the general idea, I believe we can easily compromise user parity as switching contexts too often might lead to running the intended commands in the wrong clusters. Additionally, I see this PR tries to move the `Kubeconfig` to a folder based on the cluster's name/ID? which is generally against k8s best practices.
> 
> I'd rather login to a particular cluster and logout of it before logging into another production cluster than maintaining it's context in the config. That is my sanity check for a decoupled process. We have `ocm-container` which segregates the multi-cluster login process via separate emulated environments.
> 
> @wanghaoran1988 WDYT?

Thanks for the review @supreeth7 ,
This PR mainly benefits SRE universal toolchain (#wg-sre-toolchain), which needs multi-cluster login without a containerized solution like `ocm-container` .

As I understand, multi clusters configuration allowed few methods. [K8 Ref](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/)

* Default kube config with context
* KUBECONFIG env with separate kubeconfig file

This PR uses the second method, allowing users to execute `oc` commands without specifying the context. 

If it uses the context approach, then `oc` commands may need to set the context first.
```
oc config use-context <context_nickname>
```
Please update your thoughts. 
I'm happy to change as we agreed.

File: pkg/login/kubeConfig.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+		}
+		err = f.Close()
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// set kube config env with temp kube config file
+
+	err = os.Setenv("KUBECONFIG", filename)
```
Comment (bmeng) [inline pkg/login/kubeConfig.go:57]: Maybe we can set this `KUBECONFIG` as a const?

File: pkg/login/kubeConfig.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+		kubeConfigBasePath := filepath.Join(homedir, ".kube")
+
+		return kubeConfigBasePath, nil
+	}
+
+	return kubeConfigBasePath, nil
+}
+
+func SetKubeConfigBasePath(basePath string) error {
```
Comment (bmeng) [inline pkg/login/kubeConfig.go:123]: This used in test only?

File: pkg/utils/util.go [hunk 11]
```diff
@@ -186,6 +187,19 @@ func CreateTempKubeConfig(kubeConfig *api.Config) error {
 
 }
 
+// Get default kube config
+func GetDefaultKubeConfig() api.Config {
+
```
Comment (bmeng) [inline pkg/utils/util.go:192]: Remove this blank line?

File: pkg/utils/util.go [hunk 12]
```diff
@@ -186,6 +187,19 @@ func CreateTempKubeConfig(kubeConfig *api.Config) error {
 
 }
 
+// Get default kube config
```
Comment (bmeng) [inline pkg/utils/util.go:190]: As the convention the comments should better start with the function name

File: pkg/utils/util.go [hunk 13]
```diff
@@ -186,6 +187,19 @@ func CreateTempKubeConfig(kubeConfig *api.Config) error {
...
+func GetDefaultKubeConfig() api.Config {
+
+	return defaultKubeConfig
+}
+
+// Modify Default temp kube config file name
+func ModifyTempKubeConfigFileName(fileName string) error {
```
Comment (bmeng) [inline pkg/utils/util.go:197]: As the convention the comments should better start with the function name

File: pkg/login/kubeConfig.go (hunk 10 above)
Comment (samanthajayasinghe) [reply pkg/login/kubeConfig.go:123]: Yeah. This allows modifying kube config save path, which may benefit other tools + unit tests 
basically, unit test docker containers only allow writing files to `/tmp` folder and this function helps to modify path to `/tmp` folder

File: README.md [hunk 14]
```diff
@@ -166,6 +166,24 @@ In this example, we will login to a cluster with id `123456abcdef` in production
   $ ocm backplane cloud console
   ```
 
+## Login to multiple clusters
```
Comment (supreeth7) [inline README.md:169]: This section can be moved under the Login section for a better structure

**Review by supreeth7 — CHANGES_REQUESTED** [review]

**Review by supreeth7 — APPROVED** [review]

**Pull request merged**